packages/compiler-cli/src/ngtsc/scope/src/util.ts

```diff
@@ -0,0 +1,65 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import ts from 'typescript';
+
+import {ErrorCode, makeDiagnostic, makeRelatedInformation} from '../../diagnostics';
+import {Reference} from '../../imports';
+import {ClassDeclaration} from '../../reflection';
+
+import {ComponentScopeReader} from './component_scope';
+
+export function getDiagnosticNode(
+    ref: Reference<ClassDeclaration>, rawExpr: ts.Expression|null): ts.Expression {
+  // Show the diagnostic on the node within `rawExpr` which references the declaration
+  // in question. `rawExpr` represents the raw expression from which `ref` was partially evaluated,
+  // so use that to find the right node. Note that by the type system, `rawExpr` might be `null`, so
+  // fall back on the declaration identifier in that case (even though in practice this should never
+  // happen since local NgModules always have associated expressions).
+  return rawExpr !== null ? ref.getOriginForDiagnostics(rawExpr) : ref.node.name;
+}
+
+export function makeNotStandaloneDiagnostic(
+    scopeReader: ComponentScopeReader, ref: Reference<ClassDeclaration>,
+    rawExpr: ts.Expression|null, kind: 'component'|'directive'|'pipe'): ts.Diagnostic {
+  const scope = scopeReader.getScopeForComponent(ref.node);
+
+  let relatedInformation: ts.DiagnosticRelatedInformation[]|undefined = undefined;
+  if (scope !== null) {
+    // The directive/pipe in question is declared in an NgModule. Check if it's also exported.
+    const isExported = scope.exported.directives.some(exp => exp.ref.node === ref.node) ||
+        scope.exported.pipes.some(exp => exp.ref.node);
+    if (isExported) {
+      relatedInformation = [makeRelatedInformation(
+          scope.ngModule.name,
+          `It can be imported using its NgModule '${scope.ngModule.name.text}' instead.`)];
+    } else {
+      relatedInformation = [makeRelatedInformation(
+          scope.ngModule.name,
+          `It's declared in the NgModule '${
+              scope.ngModule.name.text}', but is not exported. Consider exporting it.`)];
+    }
+  } else {
+    // TODO(alxhub): the above case handles directives/pipes in NgModules that are declared in the
+    // current compilation, but not those imported from .d.ts dependencies. We could likely scan the
+    // program here and find NgModules to suggest, to improve the error in that case.
+  }
+  let extraText: string = '';
+  if (relatedInformation === undefined) {
+    // If no contextual pointers can be provided to suggest a specific remedy, then at least tell
+    // the user broadly what they need to do.
+    extraText = ' It must be imported via an NgModule.';
+  }
+  return makeDiagnostic(
+      ErrorCode.COMPONENT_IMPORT_NOT_STANDALONE, getDiagnosticNode(ref, rawExpr),
+      `The ${kind} '${
+          ref.node.name
+              .text}' appears in 'imports', but is not standalone and cannot be imported directly.${
+          extraText}`,
```

Comment:
```suggestion
  let diagnosticMessage = `The ${kind} '${
          ref.node.name
              .text}' appears in 'imports', but is not standalone and cannot be imported directly`;
  if (relatedInformation === undefined) {
    // If no contextual pointers can be provided to suggest a specific remedy, then at least tell
    // the user broadly what they need to do.
    diagnosticMessage += ' It must be imported via an NgModule.';
  }
  return makeDiagnostic(
      ErrorCode.COMPONENT_IMPORT_NOT_STANDALONE, getDiagnosticNode(ref, rawExpr),
      diagnosticMessage,
```

Similar to the other comment, I think maybe this is a tad more straightforward.

Reply:
👍 much cleaner.